File: mazes.java
/*
 This solution uses a technique called 0/1 BFS, which finds the 
 shortest path in a graph where all edge weights are either 0
 or 1. The maze can be thought of as a graph where the nodes
 are the cells and each node has an edge to the 4 adjacent cells.
 Clearing a cell costs 1, so edges to blocked cells have weight
 1 while edges to clear cells have weight 0.
 The shortest path from S to E in this graph is equal to the
 minimum number of cells which must be cleared.
 
 0/1 BFS works the same way as regular BFS, except nodes that
 were reached over a 0 edge are added to the front of the queue
 instead of the back and their distance is not incremented. 
 This way, all possible "free" nodes are visited, whenever
 possible, before incurring any cost.
 */
import java.util.*;
public class mazes {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		
		int t = in.nextInt();
		for(int z = 1; z <= t; z ++)
		{
			//Taking in the input
			int n = in.nextInt();
			int m = in.nextInt();
			boolean[][] map = new boolean[n][m];
			int start_i = -1;
			int start_j = -1;
			int end_i = -1;
			int end_j = -1;
			
			for(int i = 0; i < n; i ++)
			{
				String row = in.next();
				for(int j = 0; j < m; j++)
				{
					char c = row.charAt(j);
					map[i][j] = c == '.' || c == 'S' || c == 'E';
					if(c=='S')
					{
						start_i = i;
						start_j = j;
					}
					if(c=='E')
					{
						end_i = i;
						end_j = j;
					}
				}
			}
			
			//stores the distance from s to each cell.
			int[][] distance = new int[n][m];
			for(int i = 0; i < n; i ++)
				Arrays.fill(distance[i],  -1);
			
			ArrayDeque<Integer> queue = new ArrayDeque<Integer>();
			
			//initially we are at cell (start_i, end_i)
			queue.add(start_i);
			queue.add(start_j);		
			distance[start_i][start_j] = 0;
	
			//These arrays store the four directions that we can move
			int[] di = new int[] {0,0,1,-1};
			int[] dj = new int[] {1,-1,0,0};
			
			while(!queue.isEmpty())
			{
				//The current cell
				int at_i = queue.poll();
				int at_j = queue.poll();
				
				//Try all 4 directions
				for(int k = 0; k < 4; k ++)
				{
					int next_i = at_i + di[k];
					int next_j = at_j + dj[k];
					
					/*
					 * If we are trying to go to a cell that is out of bounds or
					 * that has already been visited, don't.
					 */
					if(next_i >= n || next_i < 0 || next_j >= m || next_j < 0 || distance[next_i][next_j] != -1)
						continue;
					
					//If we are moving into a blocked cell
					if(!map[next_i][next_j])
					{
						//We have incurred 1 cost for clearing this cell
						distance[next_i][next_j] = distance[at_i][at_j] + 1;
						queue.add(next_i);
						queue.add(next_j);
					}
					
					//If we are moving into a clear cell
					else
					{
						//It does not cost any extra
						distance[next_i][next_j] = distance[at_i][at_j];
						
						//Add it to the front of the queue
						queue.addFirst(next_j);
						queue.addFirst(next_i);
					}
				}
			}
			
			//The answer is the distance from S to E
			System.out.println("Maze #" + z + ": " + distance[end_i][end_j]);
		}
	}
}

File: mazes.cpp
/* Eric Ly
 * Overgrown Mazes
 * HSPT 2016 Online Edition
 */

#include <iostream>
#include <algorithm>
#include <deque>
#include <cstdio>

using namespace std;

// Make a direction array
int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};
int oo = -1;
int r, c;

int main()
{
   int T; cin >> T;
   for(int cc = 1; cc <= T; cc++)
   {
      cin >> r >> c;
      string g[r];
      for(int i = 0; i < r; i++) cin >> g[i];
      int sx = 0;
      int sy = 0;
      int ex = 0;
      int ey = 0;
      int d[r][c];
      for(int i = 0; i < r; i++)
      {
         for(int j = 0; j < c; j++)
         {
            // Initialize array with infinities indicating we have not went to this location
            d[i][j] = oo;

            // If we find the start or end, save these locations
            if(g[i][j] == 'S')
            {
               sx = i;
               sy = j;
            }
            if(g[i][j] == 'E')
            {
               ex = i;
               ey = j;
            }
         }
      }

      // Make a Deque so we can add to either the front or back
      deque<int> q;
      q.push_back(sx);
      q.push_back(sy);

      // Initialize the start with a zero
      d[sx][sy] = 0;

      // We will perform a 0/1 BFS
      // Basically if we take an edge that adds a cost to our current answer to push to the back of the que
      // If we take an edge that has no cost then we can push it to the front of the que
      while(q.size() > 0)
      {
         int x = q.front(); q.pop_front();
         int y = q.front(); q.pop_front();
         for(int i = 0; i < 4; i++)
         {
            int nx = x+dx[i];
            int ny = y+dy[i];
            // Check bounds
            if(0 <= nx && nx < r && 0 <= ny && ny < c && d[nx][ny] == oo)
            {
               // We hit a bush, so add 1 to our answer
               if(g[nx][ny] == '#')
               {
                  d[nx][ny] = d[x][y]+1;
                  q.push_back(nx);
                  q.push_back(ny);
               }
               // This is a free edge so we can evaluate it right away
               else
               {
                  d[nx][ny] = d[x][y];
                  q.push_front(ny);
                  q.push_front(nx);
               }
            }
         }
      }

      // Print answer
      printf("Maze #%d: %d\n", cc, d[ex][ey]);
   }
   return 0;
}
